# Agent Customization & Prompt Engineering in AutoGen v0.4

In [1]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

load_dotenv()
api_key = os.getenv('OPENAI_API_KEY')
model_client = OpenAIChatCompletionClient(model='gpt-4o', api_key=api_key)

### Agent Customization

- We can assign a role to our agent
- help in fitting agent to specific use case


In [2]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history. Provide detailed and accuragte answers about historical events,figures and timelines'
)

In [3]:
async def test_history_expert():
    result = await asssistant.run(task = 'Who was the first President of USA?')
    print(result.messages[-1].content)
    

In [4]:
await test_history_expert()

The first President of the United States was George Washington. He served as president from April 30, 1789, to March 4, 1797. Washington was unanimously elected by the Electoral College and is often referred to as the "Father of His Country" for his pivotal role in the founding of the United States. Before becoming president, he was also a leading figure in the American Revolutionary War, serving as the Commander-in-Chief of the Continental Army. His leadership and vision helped lay the foundation for the United States' system of government and the presidency itself.


The first President of the United States was George Washington. He served from 1789 to 1797. Before his presidency, Washington was well-known for his role in the American Revolutionary War as the commander-in-chief of the Continental Army. He is also noted for his presidency because he set many of the precedents for the office that are still followed today, such as limiting himself to only two terms in office. His presidency is significant for establishing numerous structures and procedures that still exist in the U.S. government today.

# Prompt Engineering

[Prompt Engineering - OpenAI ](https://platform.openai.com/docs/guides/prompt-engineering)

In [5]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history who explaines event in a storytelling style, as if narrating a historic novel. Be vivid and engaging and focus on figures and events. Limit to 50 words'
)

In [ ]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history who explaines event in a storytelling style, as if narrating a historic novel. Be vivid and engaging and focus on figures and events. Limit to 50 words'
)

In [6]:
async def test_storytelling():
    result = await asssistant.run(task = 'Tell me about something on the signing of Declaration of Independence')
    print(result.messages[-1].content)
    

In [6]:


async def test_storytelling():
    result = await asssistant.run(task = 'Who was the first President of USA?')
    print(result.messages[-1].content)
    

In [7]:
await test_storytelling()

In the sweltering summer of 1776, Philadelphia's Independence Hall buzzed with revolution. Quills danced as Jefferson scribed fiery words, while Franklin and Adams whispered strategy amidst the clamor. On July 4th, a unanimous decision ignited a new nation’s birth. Liberty was sealed with resolve, hope, and ink.


In [9]:
import asyncio
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

# Load API key from .env
load_dotenv()
api_key = os.getenv("OPENAI_API_KEY")
model_client = OpenAIChatCompletionClient(model="gpt-4o", api_key=api_key)

# Define a simple tool
def get_weather(city: str) -> str:
    return f"The weather in {city} is sunny!"  # Dummy response

# Set up the agent
agent = AssistantAgent(
    name="weather_agent",
    model_client=model_client,
    system_message="You are a weather assistant. Use the get_weather tool when asked about weather. Add narration such as precautions to take",
    tools=[get_weather],
    reflect_on_tool_use=False
)

# Run the agent
async def main():
    result = await agent.run(task="What’s the weather in New York?")
    print(result.messages[-1].content)

await main()


The weather in New York is sunny!


In [12]:
async def main():
    result = await agent.run(task="What is the capital of India?")
    print(result.messages[-1].content)

await main()


The capital of India is New Delhi. How else may I assist you today?
